Patch/1.6/service forward bug

lazan opened the pull request.

lazan committed:
config: add topology::Update to protocol documentation

lazan committed:
common: add message type count metric for send/receive

Added functionality to keep track of number of messages sent and
received for each message type.

Every time communication::device::Inbound/Outbound receives or sends a
complete-message the counter for that message type is incremented.
* if sent, sent counter is incremented
* if received, received counter is incremented

* Added request and reply message  to extract the aggregated list of
sent/received messages.
* Added a default handler for the Request to reply with the global state
  of this metric.

This makes it possible to ask any casual process for their message
count.

This can be very useful in unittest, but also to help us to produce good
admin cli info -> We can deduce stuff in-flight. Number of specific
events, and so on.

We probably want to expose this as generic cli in the `internal`
_argument group_.

lazan committed:
service: SM sends error-reply to no-reply request on timeout/core

This patch:
If the service lookup request is a `forward_request` (a lookup from the
service-forward helper) SM does not keep track of the caller.
If a timeout or a process exit occur, SM does not send error-reply (the
caller is absent).

resolves #247

AlbsAgren reviewed (approved):
Looks good, I think the message counter has the potential to be quite useful going forward

lazan committed:
cli: add cli for the generic message count

lazan committed:
log: do not throw if we can't open logfile -> error on stderr

This patch makes sure that if we can't open/reopen/relocate casual log
file we write an error to stderr. Before an exception where thrown.

We still don't check if an actual log write is successful or not. If,
for example, the logfile gets "broken", we silently drop log entries.

lazan committed:
event: make sure we try to reopen the event file if it's "broken"

Instead of exiting the event application we try to reopen the event file
if it's "broken".

If we can't and error-log is also "broken", we log to stderr.
